### PR TITLE
[Feature] Storage volume support hdfs configuration

### PR DESCRIPTION
**Pull request by TszKitLo40**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java [hunk 1]
```diff
@@ -65,7 +66,8 @@ public CloudConfiguration build(Map<String, String> properties) {
                 getOrDefault(properties, HDFS_USERNAME, HDFS_USERNAME_DEPRECATED),
                 getOrDefault(properties, HDFS_PASSWORD, HDFS_PASSWORD_DEPRECATED),
                 getOrDefault(properties, HDFS_KERBEROS_PRINCIPAL, HDFS_KERBEROS_PRINCIPAL_DEPRECATED),
-                getOrDefault(properties, HADOOP_KERBEROS_KEYTAB, HDFS_KERBEROS_KEYTAB_DEPRECATED),
+                getOrDefault(properties, HADOOP_KERBEROS_KEYTAB, HDFS_KERBEROS_KEYTAB_DEPRECATED,
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java:69]: keytab and ticket cache are two different concepts.

Comment (TszKitLo40) [reply fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java:69]: > keytab and ticket cache are two different concepts.

Do you think we need to add a new field about ticket cache?

Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java:69]: would  be better, so there are several orthogonal ways 
- no auth
- simple
- keytab file/content + principal
- ticket cache

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 26 / 26 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudCredential.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/storagevolume/StorageVolume.java|7|7|100.00%|[]|

**Review by dirtysalt — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudCredential.java|15|15|100.00%|[]|
|:large_blue_circle:|com/starrocks/storagevolume/StorageVolume.java|7|7|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 46 / 56 (82.14%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/credential/gcp/GCPCloudConfigurationProvoder.java|0|1|00.00%|[51]|
|:large_blue_circle:|com/starrocks/credential/aws/AWSCloudConfigurationProvider.java|0|1|00.00%|[107]|
|:large_blue_circle:|com/starrocks/credential/tencent/TencentCloudConfigurationProvider.java|0|1|00.00%|[46]|
|:large_blue_circle:|com/starrocks/credential/azure/AzureCloudConfigurationProvider.java|0|1|00.00%|[97]|
|:large_blue_circle:|com/starrocks/credential/aliyun/AliyunCloudConfigurationProvider.java|0|1|00.00%|[47]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java|2|3|66.67%|[87]|
|:large_blue_circle:|com/starrocks/credential/CloudConfigurationFactory.java|7|10|70.00%|[49, 63, 68]|
|:large_blue_circle:|com/starrocks/credential/hdfs/StrictHDFSCloudConfigurationProvider.java|12|13|92.31%|[48]|
|:large_blue_circle:|com/starrocks/credential/hdfs/StrictHDFSCloudCredential.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudCredential.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/storagevolume/StorageVolume.java|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 47 / 51 (92.16%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/credential/CloudConfigurationFactory.java|7|8|87.50%|[53]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudCredential.java|14|16|87.50%|[76, 79]|
|:large_blue_circle:|com/starrocks/credential/hdfs/StrictHDFSCloudConfigurationProvider.java|11|12|91.67%|[48]|
|:large_blue_circle:|com/starrocks/credential/hdfs/StrictHDFSCloudCredential.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/storagevolume/StorageVolume.java|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 47 / 51 (92.16%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/credential/CloudConfigurationFactory.java|7|8|87.50%|[50]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudCredential.java|14|16|87.50%|[76, 79]|
|:large_blue_circle:|com/starrocks/credential/hdfs/StrictHDFSCloudConfigurationProvider.java|11|12|91.67%|[48]|
|:large_blue_circle:|com/starrocks/credential/hdfs/StrictHDFSCloudCredential.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/credential/hdfs/HDFSCloudConfigurationProvider.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/storagevolume/StorageVolume.java|9|9|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by Smith-Cruise — COMMENTED** [review]
If we already know we need to create HDFS explicitly in StorageVolume, why don't `new StrictHDFSCloudCredential()` directly?

File: fe/fe-core/src/main/java/com/starrocks/credential/CloudConfigurationConstants.java [hunk 2]
```diff
@@ -98,6 +98,7 @@ public class CloudConfigurationConstants {
     public static final String HDFS_PASSWORD = "hadoop.password";
     public static final String HDFS_KERBEROS_PRINCIPAL_DEPRECATED = "kerberos_principal";
     public static final String HDFS_KERBEROS_PRINCIPAL = "hadoop.kerberos.principal";
+    public static final String HDFS_KERBEROS_TICKET_CACHE_PATH = "hadoop.security.kerberos.ticket.cache.path";
```
[the remark this answers is not in the document]
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/credential/CloudConfigurationConstants.java:101]: yes, it is. https://github.com/apache/hadoop/blob/branch-3.3.6/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/CommonConfigurationKeys.java#L343

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java [hunk 3]
```diff
@@ -23,31 +23,35 @@
 
 import java.util.Map;
 
+import static com.starrocks.credential.CloudConfigurationConstants.HDFS_AUTHENTICATION;
+
 public class HDFSCloudCredential implements CloudCredential {
-    public static final String EMPTY = "empty";
+    public static final String NO_AUTH = "no_auth";
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java:29]: is this still necessary?

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/StrictHDFSCloudConfigurationProvider.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.credential.hdfs;
+
+import com.starrocks.credential.CloudConfiguration;
+
+import java.util.Map;
+
+import static com.starrocks.credential.CloudConfigurationConstants.HADOOP_KERBEROS_KEYTAB;
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/StrictHDFSCloudConfigurationProvider.java:21]: may be can just use `com.starrocks.credential.CloudConfigurationConstants;` to save these long list of static import?

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java [hunk 5]
```diff
@@ -104,6 +103,13 @@ public FileStoreInfo toFileStoreInfo() {
         FileStoreInfo.Builder fileStore = FileStoreInfo.newBuilder();
         fileStore.setFsType(FileStoreType.HDFS);
         HDFSFileStoreInfo.Builder hdfsFileStoreInfo = HDFSFileStoreInfo.newBuilder();
+        if (!authentication.isEmpty()) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java:106]: if authentication is provided, it is must be 'simple' or 'kerberos', looks like this condition is not checked.

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/StrictHDFSCloudCredential.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Map;
+
+public class StrictHDFSCloudCredential extends HDFSCloudCredential {
+    public StrictHDFSCloudCredential(String authentication, String username, String password, String kerberosPrincipal,
+                                     String keytab, String keytabContent, Map<String, String> hadoopConfiguration) {
+        super(authentication, username, password, kerberosPrincipal, keytab, keytabContent, hadoopConfiguration);
+    }
+
+    @Override
+    public boolean validate() {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/StrictHDFSCloudCredential.java:26]: should take some check like if authentication is provided, the value must be 'simple' or 'kerberos'.

File: fe/fe-core/src/test/java/com/starrocks/catalog/ColumnTest.java [hunk 7]
```diff
@@ -344,8 +344,8 @@ public void testLscColumn() {
         TColumn t0 = f0.toThrift();
         f0.setIndexFlag(t0, Collections.singletonList(i0), bfColumns);
 
-        Assert.assertEquals(t0.has_bitmap_index, true);
-        Assert.assertEquals(t0.is_bloom_filter_column, true);
+        Assert.assertEquals(true, t0.has_bitmap_index);
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/catalog/ColumnTest.java:347]: remove these unrelated changes. make the PR clean and focus.

File: fe/fe-core/src/test/java/com/starrocks/storagevolume/StorageVolumeTest.java [hunk 8]
```diff
@@ -22,6 +22,7 @@
 import com.staros.proto.AzBlobFileStoreInfo;
 import com.staros.proto.FileStoreInfo;
 import com.staros.proto.FileStoreType;
+import com.staros.proto.HDFSFileStoreInfo;
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/storagevolume/StorageVolumeTest.java:25]: need add ut to cover the storage provider `strict mode` and default mode behavior. use UT to help explain the code design purpose.

**Comment by TszKitLo40** [comment]
> If we already know we need to create HDFS explicitly in StorageVolume, why don't `new StrictHDFSCloudCredential()` directly?

As decussed before, We can add hint for `buildCloudConfigurationForStorage` later.

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java [hunk 9]
```diff
@@ -23,31 +23,34 @@
...
+    public static final String KERBEROS_AUTH = "kerberos";
+    protected String authentication;
     private String userName;
     private String password;
-    private String krbPrincipal;
-    private String krbKeyTabFile;
-    private String krbKeyTabData;
+    private String kerberosPrincipal;
+    private String keytab;
+    private String keytabContent;
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java:36]: looks like the above three lines variable rename is not necessary, can revert back to keep this PR as focus as possible.

File: fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java [hunk 10]
```diff
@@ -65,19 +68,17 @@ public void applyToConfiguration(Configuration configuration) {
...
-                return false;
+        if (!authentication.isEmpty()) {
+            if (authentication.equals(SIMPLE_AUTH)) {
+                return true;
+            }
+            if (authentication.equals(KERBEROS_AUTH)) {
+                if (kerberosPrincipal.isEmpty()) {
+                    return false;
+                }
+                return !(keytab.isEmpty() && keytabContent.isEmpty());
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/credential/hdfs/HDFSCloudCredential.java:79]:
```suggestion
            if (SIMPLE_AUTH.equals(authentication)) {
                return true;
            }
            if (KERBEROS_AUTH.equals(authentication)) {
                if (kerberosPrincipal.isEmpty()) {
                    return false;
                }
                return !(keytab.isEmpty() && keytabContent.isEmpty());
            }
```

**Review by kevincai — APPROVED** [review]
nice!

**Review by Smith-Cruise — APPROVED** [review]
Better to add some comment to explain StrictHDFSCloudConfiguration

**Review by dirtysalt — DISMISSED** [review]

**Review by silverbullet233 — DISMISSED** [review]

**Review by kevincai — APPROVED** [review]

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Comment by TszKitLo40** [comment]
https://github.com/Mergifyio backport branch-3.2

**Comment by TszKitLo40** [comment]
https://github.com/Mergifyio backport branch-3.1
